Preliminary zoom functionality

rlskoeser opened the pull request.

gissoo commented:
@rlskoeser thank you!! Super exciting!! Everything mostly works
A couple of issues I noticed: 
- on the actual mobile (chrome) pinching (in and out) for zooming in and out doesn't work, it rather moves off the page or kind of flickers. 
- Also, on the actual mobile (chrome) and the emulator some of the selected leaves don't get centered, they remain on the edges, even when they are not on the edge of the svg. 
- Also, on the actual mobile (chrome) using the finger to move through the leaves is not smooth, takes a lot of effort 
- on the mobile emulator (safari) labels appear without a change in the zoom level, the zoom doesn't work when selecting a leaf either, the reset button only makes the labels disappear. See below
<img width="1149" alt="Screenshot 2023-04-05 at 12 43 14 PM" src="https://user-images.githubusercontent.com/44615389/230147925-2609e0d4-2520-40f0-b5ca-1d4f614e3c42.png">

quadrismegistus commented on assets/js/timetree.js:
```diff
@@ -216,17 +216,75 @@ function TreeGraph({ nodes, links, centuries }) {
+  }
+
+  function svgWidth() {
+    return isMobile() ? 800 : 1200;
```
Are these params somewhere?

rlskoeser replied:
No, I had it hard-coded in a variable and we wanted to set a different width for mobile than desktop. I could move them into a settings variable similar to what I did with the force strengths; would that be better for documentation / readability?

quadrismegistus commented on assets/js/timetree.js:
```diff
@@ -554,6 +628,19 @@ function TreeGraph({ nodes, links, centuries }) {
+  // since svg width depends on mobile/desktop, adjust on resize
+  window.onresize = () => {
+    let newWidth = svgWidth();
+    // if the width has changed, update width, min x, svg viewbox,
+    // and reposition the simulation
+    if (newWidth != width) {
+      width = newWidth;
+      min_x = -width / 2;
+      svg.attr("viewBox", [min_x, min_y, width, height]);
+      simulation.tick(); // should reflow and update everything with new sizes...
```
(Larger, indirect thought, but) I wonder if a lot of this code here could be abstracted/generalized once Lenape is done to be a template for other d3+hugo focused projects. A lot of the code seems extensible in that sense

rlskoeser commented on assets/js/timetree.js:
```diff
@@ -216,17 +216,75 @@ function TreeGraph({ nodes, links, centuries }) {
+  const mobileQuery = window.matchMedia("(max-width: 768px)");
+
+  function isMobile() {
+    return mobileQuery.matches; // true if our query matches
```
`matches` is a boolean - here's the MDN documentation I was referencing https://developer.mozilla.org/en-US/docs/Web/API/Window/matchMedia

rlskoeser commented on assets/js/timetree.js:
```diff
@@ -413,10 +471,26 @@ function TreeGraph({ nodes, links, centuries }) {
     .on("mouseover", Leaf.highlightLeaf)
     .on("mouseout", Leaf.unhighlightLeaf);
 
+  function clickLeaf(event, el) {
+    event.stopPropagation();
+    console.log(event);
+    Leaf.selectLeaf(event);
```
Yeah, I'll have to update with new tag / selection functionality you've been working on.

My sense is that the zoom logic should stay here and the leaf code shouldn't need to know about zoom behavior.

rlskoeser commented:
@gissoo @quadrismegistus this isn't fully implemented but a fresh round of review and testing would be helpful so I know what to focus on when I come back to this.

I've updated to use the latest version of the code (refactored as part of the tag interaction work), implemented centuries as a d3 axis, and implemented semantic zooming so the tree and the axis zoom separately (no longer zooming the whole svg). I have work to do on the styling; there are some weird problems related to overflow and when it should/shouldn't be displayed; centuries are not positioned or styled correctly yet. But I think the functionality is much better than before, and the zoom worked properly for me on my phone with multitouch.

quadrismegistus reviewed (approved):
Looks great! Seems to be nearly there!!

gissoo commented:
@rlskoeser this looks amazing!!!! Thank you! The zoom on all mobile emulators and actual chrome on mobile works pretty well. Sorry, I also just checked the reset zoom button, and it works perfectly. I love the semantic zooming! 

Some things don't work well and they may not be in scope for this issue but I'll just document here, 
- it's still pretty hard to select a leaf, can't seem to select anything (might be the problem you mentioned with the visual debug covering the leaves?)
- as you said the centuries aren't styled, they look so tiny
- the centuries seem to be taking a lot of space on desktop? It didn't look this way before. 
<img width="813" alt="Screenshot 2023-04-12 at 1 58 23 PM" src="https://user-images.githubusercontent.com/44615389/231543841-45aede82-b2f8-4911-a360-d1868015742d.png">

rlskoeser commented:
@gissoo I've worked on updates to address the problems you identified, and I've implemented the tag+zoom logic you proposed. I even figured out how to calculate how to zoom in on a group of tagged leaves! Please test and review.

@quadrismegistus I've added a lot of code, let me know if you see any opportunities for refactoring or spot any problems/concerns.

quadrismegistus commented on assets/js/timetree.js:
```diff
@@ -552,27 +614,101 @@ class TimeTree {
+
+      // zoom in on the tags at the calculated scale
+      this.zoomToDatum(tagCenter, scale);
     }
```
Very slick, and readable

quadrismegistus commented on assets/js/leaves.js:
```diff
@@ -38,6 +38,9 @@ function cointoss() {
   return Math.random() < 0.5;
 }
 
+const TagSelectEvent = new Event("tag-select");
+const TagDeselectEvent = new Event("tag-deselect");
```
And we need these because we want to tag selection to do more than just updateSelection (also zoom, etc)?

rlskoeser replied:
Yes. There are now zoom behaviors associated with tag interactions. Signals seemed the cleanest way to handle that. Selecting a tag will trigger a zoom out so you can see all the tags with that leaf; deselecting a tag will zoom back in of there is a leaf selected or zoom out if not.

quadrismegistus reviewed (approved):
Awesome!

gissoo commented:
@rlskoeser this looks amazing!! Thank you 
The only thing I noticed and don't know how much it matters: on the chrome mobile emulator I can't drag to scan other leaves once zoomed in. I can do it with other browsers. On the actual chrome's mobile and firefox everything works perfectly. The tag zoom logic works perfectly except when the leaves are being cut off from the svg container, hopefully we can fix that.

jhimpele commented:
see unchecked boxes...was not seeing selected leaf under panel or getting full view of tree when closing panel. (iPhone SE, smaller screen.)

rlskoeser commented:
@jhimpele thanks for testing, sorry to see that it's not working. I'm not sure how best to troubleshoot / investigate, since it's working the way I expect in the browser mobile emulators I am able to test with. Do you want to show me in person at our next meeting?

jhimpele commented:
Yes! Sounds good.




Jeffrey Himpele.
Director, VizE Lab for Ethnographic Data Visualization
Lecturer in Anthropology

vizelab.princeton.edu<http://vizelab.princeton.edu/>


On 4/21/23, 3:11 PM, "Rebecca Sutton Koeser" wrote:


@jhimpele<https://github.com/jhimpele> thanks for testing, sorry to see that it's not working. I'm not sure how best to troubleshoot / investigate, since it's working the way I expect in the browser mobile emulators I am able to test with. Do you want to show me in person at our next meeting?

—
Reply to this email directly, view it on GitHub<https://github.com/Princeton-CDH/lenape-timetree/pull/142#issuecomment-1518238291>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AWMVXZ72XETHUHELG633RI3XCLLXDANCNFSM6AAAAAAWTE5FMY>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

gissoo commented:
@rlskoeser It works perfectly now!! Tested on all browsers' mobile emulators and FF and GC browsers on mobile. The zoom level is not changing when closing a leaf panel or the intro panel. 

There is just a lot of empty space between the tree container and the panel on FF mobile emulator.
